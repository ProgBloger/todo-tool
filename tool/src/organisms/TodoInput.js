import React from "react";
import Checkbox from "../atoms/Checkbox";

const TodoInput = () => {
    return (
        <React.Fragment>
            <Checkbox />
            <span>Currently typing</span>
            <input type="text" />
        </React.Fragment>
    )
}

export default TodoInput;