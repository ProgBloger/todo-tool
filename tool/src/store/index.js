import {combineReducers, configureStore} from "@reduxjs/toolkit";
import todoListSlice from "./todo-list-slice";

var persistedState = JSON.parse(localStorage.getItem("state"));

const store = configureStore({
    reducer: {
        todoList: todoListSlice.reducer,
    },
    preloadedState: persistedState === null ? {} : persistedState
});

store.subscribe(() => {
    var state = store.getState();

    var jsonState = JSON.stringify(state);
    localStorage.setItem("state", jsonState);
});

export default store;

export const setupStore = preloadedState => {
    return configureStore({
        reducer: {
            todoList: todoListSlice.reducer,
        },
        preloadedState: preloadedState
    })
}