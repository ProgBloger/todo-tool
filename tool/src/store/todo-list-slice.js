import { createSlice } from "@reduxjs/toolkit";

const initialState = {
    items: [],
    itemsToShow:[],
    itemsTotal: 0,
    itemsTotalActive: 0,
    // true - completed
    // false - active
    // null - all
    filter: null,
};

const filterProcessor = (itemsList, filterParam) => {
    if(filterParam == null)
        return itemsList;
    return itemsList.filter(item => item.completed == filterParam)
}

const todoListSlice = createSlice({
    name: 'todoList',
    initialState,
    reducers: {
        addItem(state, action) {
            state.itemsTotal = state.itemsTotal + 1;
            state.itemsTotalActive = state.itemsTotalActive + 1;
            console.log(action);
            const newItem = {
                id: state.itemsTotal,
                text: action.payload.text,
                completed: false,
            }

            state.items.push(newItem);

            // update the show list based on active filter
            state.itemsToShow = filterProcessor(state.items, state.filter);
        },
        removeItem(state, action) {
            const id = action.payload.id;
            const existingItem = state.items.find(item => item.id === id);
            
            // updating totals on item being removed
            state.itemsTotal--;
            if(!existingItem.completed) {
                state.itemsTotalActive--;
            }

            state.items = state.items.filter(item => item.id !== id);

            // update the show list based on active filter
            state.itemsToShow = filterProcessor(state.items, state.filter);
        },
        changeItemState(state, action) {
            // find item from the state list
            const newItemState = action.payload;
            const foundItem = state.items.find((item) => item.id === newItemState.id);
            
            // set item's state from the payload
            foundItem.completed = newItemState.completed;
            
            if(foundItem.completed) {
                state.itemsTotalActive--;
            } else {
                state.itemsTotalActive++;
            }
            
            // update the show list based on active filter
            state.itemsToShow = filterProcessor(state.items, state.filter);
        },
        fiterItems(state, action) {
            state.filter = action.payload
            state.itemsToShow = filterProcessor(state.items, state.filter);
        },
        clearCompleted(state, action){},
    },
});

export const todoActions = todoListSlice.actions;
export default todoListSlice;

