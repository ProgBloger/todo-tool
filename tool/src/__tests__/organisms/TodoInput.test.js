import renderer from 'react-test-renderer';
import expect from 'expect';
import TodoInput from '../../organisms/TodoInput';
import React from 'react';
import store from '../../store';
import { Provider } from 'react-redux';

it('renders', () => {
    // Arrange
    // Act
    var tree = renderer.create(
        <Provider store={store}>
        <TodoInput />
        </Provider>
    ).toJSON();

    // Assert
    expect(tree).toMatchSnapshot();
})