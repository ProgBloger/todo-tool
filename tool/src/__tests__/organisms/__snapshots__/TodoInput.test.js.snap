// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div
  className="container"
>
  <div
    className="inactive round"
    onClick={[Function]}
  >
    <input
      id={-1}
      type="checkbox"
    />
    <label
      htmlFor="checkbox"
    />
  </div>
  <input
    className="note"
    onKeyDown={[Function]}
    placeholder="If you want God to laugh tell him your plans"
    type="text"
  />
</div>
`;
